mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala

```diff
@@ -1142,4 +1144,47 @@ private[spark] object ReadWriteUtils {
       spark.read.parquet(path).as[T].collect()
     }
   }
+
+  def saveDataFrame(path: String, df: DataFrame): Unit = {
+    if (localSavingModeState.get()) {
+      val filePath = Paths.get(path)
+      Files.createDirectories(filePath.getParent)
+
+      Using.resource(
+        new ObjectOutputStream(new BufferedOutputStream(new FileOutputStream(filePath.toFile)))
+      ) { oos =>
+        val schema: StructType = df.schema
+        oos.writeObject(schema)
+        val it = df.toLocalIterator()
+        while (it.hasNext) {
+          oos.writeBoolean(true) // hasNext = True
+          val row: Row = it.next()
+          oos.writeObject(row)
+        }
+        oos.writeBoolean(false) // hasNext = False
+      }
+    } else {
+      df.write.parquet(path)
+    }
+  }
+
+  def loadDataFrame(path: String, spark: SparkSession): DataFrame = {
+    if (localSavingModeState.get()) {
+      Using.resource(
+        new ObjectInputStream(new BufferedInputStream(new FileInputStream(path)))
+      ) { ois =>
+        val schema = ois.readObject().asInstanceOf[StructType]
```

Comment:
This uses Java deserializer which seems **unsafe**  (risk of Remote Code Execution)

Related commit: https://github.com/apache/spark/pull/50922

Reply:
resort to arrow format suggested by @cloud-fan